backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java

```diff
@@ -0,0 +1,666 @@
+              .append(" expr ")
+              .append("SET expr.")
+              .append(ModelDBConstants.DELETED)
+              .append(" = :deleted ")
+              .append(" WHERE expr.")
+              .append(ModelDBConstants.PROJECT_ID)
+              .append(" IN (:projectIds)")
+              .toString();
+      Query deletedExperimentRunQuery =
+          session.createQuery(updateDeletedStatusExperimentRunQueryString);
+      deletedExperimentRunQuery.setParameter("deleted", true);
+      deletedExperimentRunQuery.setParameter("projectIds", projectIds);
+      deletedExperimentRunQuery.executeUpdate();
+
+      for (ProjectEntity projectEntity : projectEntities) {
+        session.delete(projectEntity);
+      }
+      transaction.commit();
+    }
+
+    LOGGER.debug("Project Deleted successfully : Deleted projects count {}", projectIds.size());
+  }
+
+  private void getRoleBindingsOfAccessibleProjects(
+      List<ProjectEntity> allowedProjects, List<String> roleBindingNames) {
+    UserInfo unsignedUser = authService.getUnsignedUser();
+    for (ProjectEntity project : allowedProjects) {
+      String projectId = project.getId();
+
+      String ownerRoleBindingName =
+          roleService.buildRoleBindingName(
+              ModelDBConstants.ROLE_PROJECT_OWNER,
+              project.getId(),
+              project.getOwner(),
+              ModelResourceEnum.ModelDBServiceResourceTypes.PROJECT.name());
+      if (ownerRoleBindingName != null) {
+        roleBindingNames.add(ownerRoleBindingName);
+      }
+
+      if (project.getProject_visibility() == ProjectVisibility.PUBLIC.getNumber()) {
+        String publicReadRoleBindingName =
+            roleService.buildRoleBindingName(
+                ModelDBConstants.ROLE_PROJECT_PUBLIC_READ,
+                projectId,
+                authService.getVertaIdFromUserInfo(unsignedUser),
+                ModelResourceEnum.ModelDBServiceResourceTypes.PROJECT.name());
+        if (publicReadRoleBindingName != null) {
+          roleBindingNames.add(publicReadRoleBindingName);
+        }
+      }
+
+      // Delete workspace based roleBindings
+      List<String> workspaceRoleBindingNames =
+          roleService.getWorkspaceRoleBindings(
+              project.getWorkspace(),
+              WorkspaceTypeEnum.WorkspaceType.forNumber(project.getWorkspace_type()),
+              projectId,
+              ModelDBConstants.ROLE_PROJECT_ADMIN,
+              ModelResourceEnum.ModelDBServiceResourceTypes.PROJECT,
+              ProjectVisibility.forNumber(project.getProject_visibility())
+                  .equals(ProjectVisibility.ORG_SCOPED_PUBLIC),
+              "_GLOBAL_SHARING");
+      roleBindingNames.addAll(workspaceRoleBindingNames);
+    }
+
+    roleService.deleteAllResources(
+        allowedProjects.stream().map(ProjectEntity::getId).collect(Collectors.toList()),
+        ModelResourceEnum.ModelDBServiceResourceTypes.PROJECT);
```

Comment:
Move this somewhere else. This is very sneaky by calling a major delete inside a function named `getRoleBindingsOfAccessibleProjects`. It should only *get* role bindings.

Reply:
Created separate functions called `deleteROleBindingsFor*` and moved all delete role bindings code in that function for each entity.